File: js/stage3/S3_PIPE_01.js
// S3_PIPE_01 — pipe (left to right)
const pipe =
  (...fns) =>
  (value) =>
    fns.reduce((acc, fn) => fn(acc), value);
// Test
const add1 = (x) => x + 1;
const double = (x) => x * 2;
const fn = pipe(add1, double);
console.log(fn(3)); // (3 + 1) * 2 = 8

File: Python/stage3/S3_PIPE_01.py
# S3_PIPE_01 — pipe: left-to-right function composition

def pipe(*fns):

    def runner(x):

        for fn in fns:

            x = fn(x)

        return x

    return runner


# Tests

f = lambda x: x + 1

g = lambda x: x * 2

h = lambda x: x - 3

p = pipe(f, g, h)

print(p(10))  # ((10 + 1) * 2) - 3 = 19
 